ipa-core/src/protocol/dp/distributions.rs

```diff
@@ -125,4 +216,191 @@ mod tests {
+    }
+    #[test]
+    fn test_truncated_double_geometric_hoffding() {
+        assert!(test_internal_generate_truncated_double_geometric_hoffding());
+    }
+    fn test_internal_generate_truncated_double_geometric_hoffding() -> bool {
+        let mut rng = rand::thread_rng();
+        let number_samples = 1000;
+        let failure_prob = 1.0;
+        let s = 1.0; // Set s to some value (e.g., 1.0)
+        let n = 25; // Set n to some value (e.g., 25)
+        #[allow(clippy::cast_precision_loss)]
+        let t = f64::sqrt(
+            f64::powf(2.0 * (f64::from(n)), 2.0) / (-2.0 * (f64::from(number_samples)))
+                * f64::ln(failure_prob / 2.0),
+        );
+        println!("t: {t:?}");
+        let mut samples = Vec::new();
+        // Sample number_samples values from the generate_truncated_double_geometric function
+        let distribution = TruncatedDoubleGeometric::new(s, n)
+            .expect("Truncated Double Geometric not constructed properly");
+        for _ in 0..number_samples {
+            let sample = distribution.sample(&mut rng);
+            samples.push(sample);
+        }
+        // Compute the sample mean
+        let sample_mean = f64::from(samples.iter().sum::<u32>()) / f64::from(number_samples);
+        // println!("sample_mean: {:?}", sample_mean);
+        // Check that the sample mean is within some distance of the expected value
+        let expected_mean = f64::from(n);
+        // println!("expected_mean: {:?}", expected_mean);
+        (sample_mean >= expected_mean - t) && (sample_mean <= expected_mean + t)
+    }
+    #[test]
+    fn test_truncated_double_geometric_sample_dist() {
+        let mut rng = rand::thread_rng();
+        let epsilon = 1.0;
+        let s = 1.0 / epsilon;
+        let n = 25;
+        let distribution = TruncatedDoubleGeometric::new(s, n)
+            .expect("failed to construct TruncatedDoubleGeometric");
+        let num_samples = 100_000;
+        let mut samples = Vec::new();
+        // Sample 1000 values from the generate_truncated_double_geometric function
+        for _ in 0..num_samples {
+            let sample = distribution.sample(&mut rng);
+            assert!(sample <= 2 * n);
+            samples.push(sample);
+        }
+        // Compute the observed probability for each value in the range [0, 2*n)
+        let mut histogram = HashMap::new();
+        for value in samples {
+            *histogram.entry(value).or_insert(0) += 1;
+        }
+        let mut sorted_keys: Vec<u32> = histogram.keys().copied().collect();
+        sorted_keys.sort_unstable();
+        // Compute the expected probability for each value in the range [0, 2*n]
+        #[allow(clippy::cast_precision_loss)]
+        let normalizing_factor = (1.0 - E.powf(-epsilon))
+            / (1.0 + E.powf(-epsilon) - 2.0 * E.powf(-epsilon * (f64::from(n + 1)))); // 'A' in paper
```

Comment:
Lots of use of `E.powf(-epsilon)` here.  Consider creating a temporary to hold that value.

Reply:
indeed, in fact in the paper we used r = e^{-epsilon} to clean up the presentation.  don't know why I didn't keep that here